keps/sig-scheduling/4187-a-generic-scheduler-cache-extension/README.md

```diff
@@ -0,0 +1,777 @@
+```go
+// FakeExtendedResource implements the ExtendedResource
+type FakeExtendedResource struct {
+    availableCapacity int64
+}
+
+// Aggregate the resource
+func (pc *FakeExtendedResource) AddPod(pod *v1.Pod) error {
+    req := ParsePodRequest(pod)
+    if req <= 0 {
+        return err
+    }
+
+    pc. availableCapacity -= mapRequest(req) // the mapRequest function maps the req to the estimated to-be-used FakeExtendedResource
+    return nil
+}
+
+// remove the resource consumed by the pod
+func (pc *FakeExtendedResource) RemovePod(pod *v1.Pod) error {
+    req := ParsePodRequest(pod)
+    if req <= 0 {
+        return err
+    }
+
+    pc. availableCapacity += mapRequest(req)
+    return nil
+}
+```
+
+- Implement the ExtendedResource.Handle which is used to keep the resource's allocable resource update-to-date. Custom Resource(CR) is one of the methods to update the resource allocable resource by a third-party agent or a system admin. And the Handle watches the resource changes and updates the extended resource accordingly. The example uses a self-defined CR, and the CR includes two fields: nodeName and fakeResource. The fakeResource stands for the actual allocable fake resource and nodeName stands for the name of the specific node. 
+``` go
+type Handle struct {
+    cache cache.ExtendedCache
+}
+
+func (fh *Handle) Run(c cache.ExtendedCache, f informers.SharedInformerFactory) error {
+
+    fh.cache = c
+    client, err := versioned.NewForConfig(config)
+    fakeResourceInformerFactory := externalin.NewSharedInformerFactory(client, 0)
+
+    nodeInformer := fakeResourceInformerFactory.Ioi().V1().NodeFakeResourceStatuses().Informer()
+
+    nodeInformer.AddEventHandler(clientcache.ResourceEventHandler{
+        Handler: clientcache.ResourceEventHandlerFuncs{
+            AddFunc:    fh.addResource,
+            UpdateFunc: fh.updateResource,
+            DeleteFunc: fh.deleteResource,
+        },
+    })
+}
+
+func (fh *Handle) updateResource(oldObj, newObj interface{}) {
+    …
+    oldVal := old.Spec.FakeResource
+    newVal := new.Spec.FakeResource
+
+    if oldVal != newVal {
+        res := fh.cache.GetExtendedResource(new.Spec.NodeName)
+        res.(*FakeExtendedResource).SetAvail(newVal)
+        fh.cache.SetExtendedResource(new.Spec.NodeName, res)
+    }
+}
+```
+
+#### Example Implementation of a Scheduler Plugin
+
+A scheduler plugin is to be implemented to initialize the extended resource handles, register them in the registry and fetch the extended resources through the handles at these hook points.
+``` go
+// New initializes a new plugin and returns it.
+func New(_ runtime.Object, h framework.Handle) (framework.Plugin, error) {
+    // initialize the extended resource handles and register them in the registry
+   extendedResourceHandle, err := h.ExtendedResourceHandleRegistry().Register(ResourceType, func() (extendedresource.Handle, error) {
+        return &FakeResourceHandle{}, nil
+    })
```

Comment:
Does this change only intend for out-of-tree plugin? and is there any existing out-of-tree plugin impacted by this change?

Reply:
The change is intended for but not limit to out-of-tree plugins (e.g. in-tree plugin may also uses it). There is no impact to existing out-of-tree plugins as they do not use the cache extension at all.